Update to bsb_helper (PR 2/3)

bsansouci opened the pull request:
This PR aims at turning bsb_helper into a packer / linker.

This is kind of a big change and is open to criticism :)

I added the following flags to `bsb_helper`:
- `-MD-bytecode path/to/input.re` / `-MD-native path/to/input.re` those are useful when generating the `.d` file for compiling to `cmo` or `cmx` files (see the changes in `depends/depends_post_process.ml` which should explain)
- `-bs-main MyModule` used only for linking. It crawls the modules in the current project starting by that given one and lists them in the right order. This is used to figure out in what order to list the `cmo` files when linking
- `-I` this allows us to tell bsb_helper about where it can find precompiled `lib.a` files (for external dependencies basically)
- `-link-bytecode path/to/file.byte` / `-link-native path/to/file.native` tells bsb_helper that we're linking either byte code or native and what the output file should be called
- `-pack-bytecode` / `-pack-native` tells bsb_helper that we're packing a library file

The code is then pretty straightforward. For linking, for each file we look at the header part of the mlast file to get the set of dependencies of that file. We build a dependency graph with all of those. We crawl the graph from the main module to build an ordered list of deps. We check if you used any library from `otherlibs` (Unix, Bigarray etc...) and auto-include those. Then we just call the compiler with the right args.
For packing we don't need to crawl, we just use pack all of the files without caring about which ones are used. This might be a concern later. I think the ocaml linker is smart enough to drop everything that's not used.

It also adds a bunch of literals that we'll find useful in the future.

bobzhang commented:
@bsansouci I am in the process of refactoring to get rid of removing recursive calls to bsb #1412 , once it is done, I will review this PR, sorry for the delay

bsansouci commented:
No problem! I'll update this PR if needed once your chang is in :)

bobzhang commented:
@bsansouci it wold be nice to avoid pull in compiler-libs in bsb_helper utility

bsansouci commented:
Ok done. 

I also moved the code I wrote into a file called bsb_helper_extract. We talked about removing deps more generally but I'm not entirely sure if it makes sense for `ast_extract`, since it uses `bs_exception` which uses Location from `compiler-libs`. I'm not 100% sure what `Location` does here, but if this means we can have better / more precise error messages then I think we should keep that. If you think it's not totally necessary I could totally fix that up and have ast_extract not depend on compiler-libs and then move the code I wrote back into there.

I think it might not be totally necessary as the code I wrote does behave slightly differently in subtle ways (different from what already exists in ast_extract). Unless you can think of a simple way to unify some functions. I think it could be possible but it might make the whole function more convoluted. I wanted to aim for simplicity really.

bobzhang commented on jscomp/bin/whole_compiler.ml:
```diff
@@ -84497,12 +84510,12 @@ and statement_desc top cxt f (s : J.statement_desc) : Ext_pp_scope.t =
         statement false cxt f nest 
       | Some s2 -> 
-        P.newline f;
+        P.space f;
```
can you do a rebase?

bobzhang commented on jscomp/bsb/bsb_helper_extract.ml:
```diff
@@ -0,0 +1,88 @@
+    && String_map.mem current ast_table then
+      begin
+        String_set.iter
+          (visit
```
when you add a new `.ml` file, it would be nice to add `.mli` as well

bsansouci replied:
Done.

bobzhang commented on jscomp/bsb/bsb_helper_main.ml:
```diff
@@ -22,10 +22,44 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
 
+let cwd = Sys.getcwd ()
```
`cwd` is a lazy and stateful variable, can we make it an argument

bsansouci replied:
Done.

bobzhang commented on jscomp/depends/depends_post_process.ml:
```diff
@@ -51,21 +52,29 @@ let read_deps fn =
+  ~compilation_kind
+  (fn : string)
+  index : unit =
+  let suffix_cmj_or_cmo = match compilation_kind with
```
the name is misleading, `cmi,x,o`

bsansouci replied:
Done.

bobzhang commented on jscomp/bsb/bsb_helper_main.ml:
```diff
@@ -34,7 +68,154 @@ let () =
+      String_set.add ("bigarray" ^ file_extension) acc
+    else if String_set.mem "Str" v then
+      String_set.add ("str" ^ file_extension) acc
+    else if String_set.mem "Num" v then
```
can we move this function into a separate file

bsansouci replied:
Done

bobzhang commented on jscomp/bsb/bsb_helper_main.ml:
```diff
@@ -34,7 +68,154 @@ let () =
 
-    ] annoymous usage
+let () = match !pack_flag with
+| Some pack_byte_or_native ->
```
per the discussion, we are not going to support -pack, no?

bsansouci replied:
This isn't for -pack, sorry if this is misleading.
This is for packing an external dependency into a `.a` or `.cmxa` file. Does it make sense?

bobzhang commented on jscomp/bsb/bsb_helper_main.ml:
```diff
@@ -34,7 +68,154 @@ let () =
+ * *)
+let () = match !link_flag with
+| Some link_byte_or_native ->
+  begin match !main_module with
```
can we wrap it as a function instead of always executing such code path?

bobzhang replied:
better to be in a separate module

bsansouci replied:
Done.

bsansouci commented:
Ok I think I fixed everything and added the files in bin (I didn't before just to make the PR easier to read :) )

bobzhang commented on jscomp/bin/bsppx.ml:
```diff
@@ -2364,7 +2371,7 @@ let report_error ppf err =
 
 let error_of_printer loc print x =
-  errorf ~loc "%a@?" print x
+  errorf_prefixed ~loc "%a@?" print x
```
hi, you need update the OCaml submodule, we cherry-picked a commit from upstream

bsansouci replied:
Done.

bobzhang commented on jscomp/bsb/bsb_helper_dep_graph.mli:
```diff
@@ -0,0 +1,32 @@
+val sort_files_by_dependencies : domain:String_set.t -> String_set.t String_map.t -> string Queue.t
+
+val simple_collect_from_main :
+           ?alias_map:String_set.elt String_hashtbl.t ->
```
could you add some brief comments for each function, also it would be nice that only needed API is exposed.
This seems to be duplicate of some modules we have now, is it possible to make it reusable?

bsansouci replied:
Mmmh those are similar to what exist in `ast_extract`.

`sort_files_by_dependencies` was copied over and changed to remove the dependency on `bs_exception` which in turn depends on `compiler-libs`. I'm not sure how to make it better apart from refactoring bs_exception to not depend on compiler-libs.

`simple_collect_from_main` assumes that you already have the `ast_table` and doesn't allocate records like `Ast_extract.collection_from_main`. It's the same core logic though.

bobzhang commented on jscomp/bsb/bsb_helper_extract.mli:
```diff
@@ -0,0 +1,25 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
+
+val read_dependency_graph_from_mlast_file : string -> String_set.t
```
some comments would be nice, I am thinking of that we should document such format (maybe in text mode)

bsansouci replied:
Oh the header format? Yeah it'd be nice to be able to point to at least a comment explaining the format. Do you want me to add it right here and next to the function that writes the header?

bobzhang commented on jscomp/bsb/bsb_helper_linker.mli:
```diff
@@ -0,0 +1,27 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
+
+type link_t = LinkBytecode of string | LinkNative of string
```
would link behave differently in byte code model from native model, if not, why do we need make such types?

bsansouci replied:
It won't necessarily behave differently, but it'll use a different compiler and assume different file extensions. If I called it `Bytecode` and `Native` then they would look exactly as the ones inside `bsb_helper_packer`. So I would have to put them in a different file to share them but I'm not sure where, because I don't want a cyclic dep.

bobzhang commented on jscomp/bsb/bsb_helper_packer.ml:
```diff
@@ -0,0 +1,68 @@
+    sorted_tasks in
+  (* This list will be reversed so we append the otherlibs object files at the end, and they'll end at the beginning. *)
+  let all_object_files = List.rev (list_of_object_files @ (Bsb_helper_dep_graph.get_otherlibs_dependencies dependency_graph suffix_library_files)) in
+  if List.length all_object_files > 0 then begin
```
```ocaml
all_object_files <> []
```

bsansouci replied:
Done.

bobzhang commented on jscomp/bsb/bsb_helper_main.ml:
```diff
@@ -17,24 +17,101 @@
+    " set the main entry module. Only used in conjunction with -link-bytecode and -link-native";
+
+    "-I",  (Arg.String (add_include (Sys.getcwd ()))),
+    " add dir to search path for the linker and packer";
```
why do we need this flag? also `Sys.getcwd` is always on  the code path which is not necessary for js targets

bsansouci replied:
Oh whoops sorry I thought I was fixing that.

bobzhang commented on jscomp/bsb/bsb_helper_main.ml:
```diff
@@ -17,24 +17,101 @@
+  | Some main_module ->
+    Bsb_helper_linker.link 
+      link_byte_or_native
+      ~main_module:main_module
```
any way to make sure such operation is not always on the code path?

bsansouci replied:
I'm not sure what you mean... This only happens if `link_flag` contained `Some _` right? So only if the user supplied the parameter.

bobzhang replied:
this patch looks generally not invasive, however, here such code path always gets executed worries me, you can always have a new flag to trigger it to be executed (bsb_helper cmd line is not client facing)

bsansouci commented:
Ok so I think the biggest issue here is that I have 2 functions that are very similar to what you already had in `ast_extract`. Those functions are `sort_files_by_dependencies` and `simple_collect_from_main`. The fist one I had to copy/paste because it's using `bs_exception` which depends on `compiler-libs`. The 2nd one I changed a bit because I didn't need as complex of a tree as what `Ast_extract.collect_from_main` does. Though we could merge the latter, I think it might just make things harder to read. `simple_collect_from_main` is pretty simple in terms of what it does, it starts from a module and crawls the dependency tree from there to make a topological list of modules.

An alternative would be to update `ast_extract` to not depend on compiler-libs, but that would probably be a big change, and lead to slightly worse error message (we'd have to make bs_exception not depend on `compiler-libs` or remove its usage inside `ast_extract`).

bobzhang commented on jscomp/bsb/bsb_helper_main.ml:
```diff
@@ -17,24 +17,101 @@
+  Ext_filename.normalize_absolute_path (Ext_filename.combine cwd s )
+
+let process_include cwd s =
+  let i = Ext_string.rindex_neg s '?'   in
```
`'?'` is a very special logic in bspack, I don't think it is needed here

bobzhang commented on jscomp/depends/depends_post_process.ml:
```diff
@@ -49,23 +49,32 @@ let read_deps fn =
+  index : unit = 
+  let suffix_cmixo = match compilation_kind with
+  | Js       -> Literals.suffix_cmj
+  | Bytecode -> Literals.suffix_cmo
```
should be `Literals.suffix_cmi` byte code doe not do cross module inlining

bsansouci replied:
Done.

bobzhang replied:
note bytecode module b depends on a, you only need record `b.cmo : a.cmi` but record `b.cmo : a.cmo` is not that bad (slightly slower)

bobzhang commented on jscomp/bsb/bsb_helper_dep_graph.ml:
```diff
@@ -0,0 +1,114 @@
+  let set_of_otherlib_deps = String_map.fold (fun k v acc ->
+    if String_set.mem "Unix" v then
+      String_set.add ("unix" ^ file_extension) acc
+    else if String_set.mem "Bigarray" v then
```
can we refactor such code  a bit

bsansouci replied:
I'm not totally sure how you want this refactor. I think a constant number of Set lookups might be the fastest way to do this.

bsansouci replied:
I made it look maybe nicer, tell me what you think.

bobzhang commented on jscomp/bsb/bsb_helper_dep_graph.ml:
```diff
@@ -0,0 +1,114 @@
+    else if String_set.mem current !worklist then
+      begin
+        next current |>
+        String_set.iter
```
can you mark it a duplication to some modules, so in the future, we can do refactoring to get rid of it

bsansouci replied:
Done.

bobzhang commented on jscomp/bin/bsb_helper.d:
```diff
@@ -1,19 +1,40 @@
+bin/bsb_helper.ml : ext/string_set.mli
+bin/bsb_helper.ml : common/js_config.ml
 bin/bsb_helper.ml : ext/ext_filename.ml
+bin/bsb_helper.ml : common/js_config.mli
```
hmm, `Js_config` should not be pulled in

bsansouci replied:
Done.

bobzhang commented:
this PR looks okay to me once comments are addressed

bsansouci commented:
I addressed some of your comments.
I'm still not sure how to avoid having that codepath be executed for packer/linker. I need to wait until we've processed all the arguments because you pass in an unbounded number of files at the end of the call, like `bsb_helper -link-native file1.cmx file2.cmx file3.cmx` right? I did make it slightly nicer by bringing the two flags together.

bobzhang commented on jscomp/bin/bsdep.ml:
```diff
@@ -25124,8 +25132,15 @@ let package_json = "package.json"
 let bsconfig_json = "bsconfig.json"
 let build_ninja = "build.ninja"
 
+let library_file = "lib"
```
what does this mean? some comments?

bsansouci replied:
Done.

bobzhang commented:
@bsansouci would `bsb_helper file1.cmx file2.cmx file3.cmx -link-native` work?

bsansouci commented:
Ok I fixed all of those! :)

bobzhang commented:
@bsansouci it looks good to go, can you sign your COMMIT MESSAGE (see CONTRIBUTION.md), thanks

bsansouci commented:
Done :D
